// yt-transcript/dist/index.mjs
import axios from 'axios';
import { xml2json } from 'xml-js';

function extractCaptionJson(html) {
  const splitHtml = html.split('"captions":');
  if (splitHtml.length <= 1) {
    if (html.includes('class="g-recaptcha"'))
      return [];
    else if (html.includes('"playabilityStatus":'))
      return [];
    else
      return [];
  }
  try {
    const json = JSON.parse(splitHtml[1].split(',"videoDetails')[0].replace("\n", ""));
    const playerCaptionsTracklistRenderer = json?.playerCaptionsTracklistRenderer;
    if (!playerCaptionsTracklistRenderer)
      return [];
    const captionTracks = playerCaptionsTracklistRenderer?.captionTracks;
    if (!captionTracks)
      return [];
    return captionTracks;
  } catch (error) {
    return [];
  }
}

function toNumber(value) {
  return Number(value);
}

async function fetchTranscript(url) {
  const response = await axios.get(url, {});
  const xml = response.data;
  const json = covertXmlToJson(xml);
  return JSON.parse(json);
}
function covertXmlToJson(xml) {
  return xml2json(xml, { compact: true, spaces: 4 });
}
function formatTranscript({ transcript }) {
  const result = [];
  const text = transcript.text;
  text.forEach((t) => {
    result.push({
      start: toNumber(t._attributes.start),
      duration: toNumber(t._attributes.dur),
      text: t._text
    });
  });
  return result;
}

async function getAllTranscripts$1(html) {
  return extractCaptionJson(html);
}
function findTranscriptByLanguage$1(transcripts, languageCode) {
  return transcripts.find((t) => t.languageCode === languageCode) || null;
}
async function getTranscript$1(html, languageCode) {
  let transcript = null;
  const transcripts = await getAllTranscripts$1(html);
  if (transcripts.length === 0)
    return null;
  if (languageCode) {
    transcript = findTranscriptByLanguage$1(transcripts, languageCode);
    if (!transcript)
      return null;
  } else {
    transcript = transcripts[0];
  }
  const url = transcript.baseUrl;
  const data = await fetchTranscript(url);
  const formatted = formatTranscript(data);
  return formatted;
}

var __defProp$1 = Object.defineProperty;
var __defNormalProp$1 = (obj, key, value) => key in obj ? __defProp$1(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField$1 = (obj, key, value) => {
  __defNormalProp$1(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
class YtTranscriptChrome {
  /**
   * Create a new instance of YtTranscript
   * @param {html} param0
   */
  constructor({ html }) {
    __publicField$1(this, "html");
    this.html = html;
  }
  /**
   * List all available transcripts for the video
   * @returns {Promise<TrasncriptInfo[]>} returns a list of available transcripts
   */
  async listAllTranscripts() {
    return await getAllTranscripts$1(this.html);
  }
  /**
   * Get transcript by language
   * @param {string} languageCode
   * @returns {Promise<TrasncriptInfo | null>} returns a transcript object or null if not found
   */
  async getTranscriptByLanguage(languageCode) {
    const transcripts = await this.listAllTranscripts();
    return findTranscriptByLanguage$1(transcripts, languageCode);
  }
  /**
   * Get transcript by language
   * @param {string} languageCode
   * @returns {Promise<YTranscript | null>} returns a transcript object or null if not found
   */
  async getTranscript(languageCode) {
    return await getTranscript$1(this.html, languageCode);
  }
}

const WATCH_URL = "https://www.youtube.com/watch?v=";
async function _getYTHtml(videoId, options) {
  const url = `${WATCH_URL}${videoId}`;
  const response = await axios.get(url, {
    headers: {
      "Accept-Language": options?.languageCode || "en-US",
      "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36",
      ...options?.headers
    }
  });
  return response.data.replace(/&quot;/g, '"').replace(/&#39;/g, "'").replace(/&#x3A;/g, ":").replace(/&lt;/g, "<").replace(/&gt;/g, ">").replace(/&amp;/g, "&");
}
async function fetchVideoHtml(videoId, options) {
  let html = await _getYTHtml(videoId, options);
  if (html.includes('action="https://consent.youtube.com/s')) {
    const match = html.match(/name="v" value="(.*?)"/);
    if (!match)
      throw new Error("Failed to find video id");
    const consent = `YES+${match[1]}`;
    const cookie = `CONSENT=${consent}; domain=.youtube.com; path=/;`;
    const headers = {
      Cookie: cookie
    };
    html = await _getYTHtml(videoId, {
      headers: {
        ...options?.headers,
        ...headers
      },
      languageCode: options?.languageCode
    });
  }
  return html;
}

async function getAllTranscripts(videoId, options) {
  const html = await fetchVideoHtml(videoId, options);
  return extractCaptionJson(html);
}
function findTranscriptByLanguage(transcripts, languageCode) {
  return transcripts.find((t) => t.languageCode === languageCode) || null;
}
async function getTranscript(videoId, languageCode) {
  let transcript = null;
  const transcripts = await getAllTranscripts(videoId, { languageCode });
  if (transcripts.length === 0)
    return null;
  if (languageCode) {
    transcript = findTranscriptByLanguage(transcripts, languageCode);
    if (!transcript)
      return null;
  } else {
    transcript = transcripts[0];
  }
  const url = transcript.baseUrl;
  const data = await fetchTranscript(url);
  const formatted = formatTranscript(data);
  return formatted;
}

function extractVideoId(url) {
  const regex = /(?:https?:\/\/)?(?:www\.)?(?:youtube\.com\/(?:[^/]+\/.+\/|(?:v|e(?:mbed)?)\/|.*[?&]v=)|youtu\.be\/)([^"&?/ ]{11})/i;
  const match = url.match(regex);
  return match ? match[1] : null;
}

var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
class YtTranscript {
  /**
   * Create a new instance of YtTranscript
   * @param {YtTranscriptConstructor} param0
   */
  constructor({ url, videoId }) {
    __publicField(this, "videoId");
    if (url) {
      const urlVideoId = extractVideoId(url);
      if (!urlVideoId)
        throw new Error("Invalid YouTube url");
      this.videoId = urlVideoId;
    } else if (videoId) {
      this.videoId = videoId;
    } else {
      throw new Error("Need to provide url or videoId");
    }
  }
  /**
   * List all available transcripts for the video
   * @returns {Promise<TrasncriptInfo[]>} returns a list of available transcripts
   */
  async listAllTranscripts() {
    return await getAllTranscripts(this.videoId);
  }
  /**
   * Get transcript by language
   * @param {string} languageCode
   * @returns {Promise<TrasncriptInfo | null>} returns a transcript object or null if not found
   */
  async getTranscriptByLanguage(languageCode) {
    const transcripts = await this.listAllTranscripts();
    return findTranscriptByLanguage(transcripts, languageCode);
  }
  /**
   * Get transcript by language
   * @param {string} languageCode
   * @returns {Promise<YTranscript | null>} returns a transcript object or null if not found
   */
  async getTranscript(languageCode) {
    return await getTranscript(this.videoId, languageCode);
  }
}

export { YtTranscript, YtTranscriptChrome };
